socketcan - added loopback and kernel time-stamping

TomDeRybel opened the pull request:
Hello Pavel,

As discussed, here are the changes. There are a few TODO (questions) left for you. Thus far, this seems to work well. 

While I was as it, I've also ported the kernel timestamping code from uavcan to this driver, to resolve the pre-existing todo and anticipate the timing requirements for my application.

As per usual, all comments welcome.

Best regards,
Tom

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+
+        // Determine the size of the ancillary data and zero-initialize the buffer for it.
+        // We require space for both the receive message and the time stamp.
+        // TODO: Aligned storage is used. --> Don't know how to do this in a portable way....
+        // TODO: C complains that the control_size is not known at compile time, hence the use of malloc/free.
+        //       Should not be needed, as those structs are fixed-size?? Ah, cmsghdr has a flexarray extension in c99?
```
Yep. Why don't you just follow the recommendation given in the docs here: https://man7.org/linux/man-pages/man3/cmsg.3.html#EXAMPLES (tl;dr: use union).

TomDeRybel replied:
Learned something new. Also, turns-out older manpages don't have this detailed example.

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
-            return getNegatedErrno();
+            const int neg_errno = getNegatedErrno();
+
+            // Return either the negated error code, or zero in case the socket is marked nonblocking and the receive
+            // operation would block. POSIX.1-2001 allows either EWOULDBLOCK or EAGAIN to be returned for this case,
+            // and does not require these constants to have the same value. Thus, we check for both possibilities.
+            free(control);
+            return (read_size < 0 && (neg_errno == -EWOULDBLOCK || neg_errno == -EAGAIN)) ? 0 : neg_errno;
```
If the `poll(...)` says that there is data waiting to be read but `recvmsg(...)` says EWOULDBLOCK/EAGAIN, you have something sketchy going on and you should perhaps return an error. In the original code, it is handled by the next condition right below this one. I recommend you replace `if (read_size <= 0)` with `if (read_size < 0)` (strictly less) and remove the comment about the non-blocking case.

TomDeRybel replied:
Reviewing the logic, your proposal makes sense. Adjusted accordingly.

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
             return 0;  // Not an extended data frame -- drop silently and return early.
         }
 
+        const bool loopback_frame = (msg.msg_flags & (int) MSG_CONFIRM) != 0;  // NOLINT
```
Bitwise operations on negatives are implementation-defined. Some static analyzers complain when you perform bitwise operations on signed integers for this reason.

```suggestion
        const bool loopback_frame = ((uint32_t) msg.msg_flags & (uint32_t) MSG_CONFIRM) != 0;  // NOLINT
```

TomDeRybel replied:
Gotcha.

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+        // Obtain the CAN frame TAI time stamp from the kernel.
+        // Comparing to clock_gettime(CLOCK_TAI, &ts) proved that the time stamp is from the desired, monotonous
+        // TAI clock source.
+        // TODO: should I keep these assertion checks?
```
Yes.

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+        {
+            assert(0);
+            free(control);
+            return -1;
```
```suggestion
            return -EIO;
```

pavel-kirienko commented on socketcan/libcanard/src/socketcan.c:
```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+        }
+
+        // Obtain the CAN frame TAI time stamp from the kernel.
+        // Comparing to clock_gettime(CLOCK_TAI, &ts) proved that the time stamp is from the desired, monotonous
+        // TAI clock source.
```
TAI is not monotonous -- it may jump and change rate to account for the local clock error. More importantly, the timestamps supplied by the kernel are, in fact, CLOCK_REALTIME, so you need to update the docs accordingly, replacing CLOCK_TAI with CLOCK_REALTIME. The kernel/POSIX documentation is a bit lacking here, unfortunately. https://stackoverflow.com/questions/60971556/which-clock-does-so-timestamp-socket-option-use

TomDeRybel replied:
Thanks for the clarification. I've read the post, followed some of the links therein, and update accordingly.
I'm still left with a couple questions regarding this (and an uneasy feeling):

- I just effectively changed the API behaviour from CLOCK_TAI to CLOCK_REALTIME in a way that won't be evident to someone just updating this file in their code base.

- In the UAVCAN spec, TAI is the preferred time system. Is there a way to convert REALTIME to TAI without silly overhead? Mainly, I have an instinctive unease about mixing time systems, and strive to keep things uniform (rule of least astonishment, and I am easily astonisch).

- This timestamping clock chosen in the kernel makes me wonder what will happen when a leap second occurs. Suddenly, CAN frames with a tight deadline might be considered expired when the nodes are still updating their clocks (because time sync is not instantaneous), yet the time providing node has already applied the leap second. (In my project, the time provider is the same as the main data processor). It seems to me that, right at this junction, all these tight deadline frames will be rejected until the node clocks have synced-up.

pavel-kirienko replied:
> I just effectively changed the API behaviour from CLOCK_TAI to CLOCK_REALTIME in a way that won't be evident to someone just updating this file in their code base.

This is not quite a library, it's more of an example, but in general, you are right, we should add a changelog. Can you please add one in the header comment at `socketcan.h`?

> In the UAVCAN spec, TAI is the preferred time system. Is there a way to convert REALTIME to TAI without silly overhead? Mainly, I have an instinctive unease about mixing time systems, and strive to keep things uniform (rule of least astonishment, and I am easily astonished).

I am not sure. A long time ago I asked a very similar question: "how do I convert between CLOCK_MONOTONIC and CLOCK_REALTIME?" My research led me exactly nowhere and on Stackoverflow I only got stupid advice to query both and subtract one from the other, so I ended up writing a user-space estimator using the Olson time lag recovery algorithm. It was not fun. In this case it may be much easier because as long as the number of leap seconds does not change, the difference is constant, and it is always an integer number of seconds, so the silly method of querying both and subtracting might actually work (proper precautions against non-uniform read latency assumed).

Mixing time systems really does create confusion and you may shoot yourself in the foot easily but it doesn't make the code non-compliant so it's not a catastrophe.

> This timestamping clock chosen in the kernel makes me wonder what will happen when a leap second occurs. Suddenly, CAN frames with a tight deadline might be considered expired when the nodes are still updating their clocks (because time sync is not instantaneous), yet the time providing node has already applied the leap second. (In my project, the time provider is the same as the main data processor). It seems to me that, right at this junction, all these tight deadline frames will be rejected until the node clocks have synced-up.

Yes, this is why one should steer clear from CLOCK_REALTIME and UTC clock in general when real-time systems are concerned, but the Linux kernel gives us no choice. There are some proposals floating to abolish the leap second, by the way.

TomDeRybel replied:
I'll add the changelog.

This thus all seems like there is no reasonably ready solution. The query and subtract approach indeed doesn't pass the sniff test in this context. Doing it right is also very error prone. Seems that validation to the monotonic clock would be required to make sure that the sampled (UTC) times are right and no one just happened to have adjusted the clock an hour backward or any other such thing. Or there was no sudden re-sync to GPS on a smaller but still relevant time-scale that would be, essentially, hard/impossible to detect.

I recall interesting incidents, such as a military aircraft passing the international dateline the wrong way round and, well, you can guess what happened next... Once, I read part of a design document regarding C coding for such systems and there was a blanket ban on using anything to do with ctime.

A conclusion from all of this could be that could be to keep the time node clock in UTC (to avoid daylight savings) and just accept that every so often, during X ms while the time synchronization system updates these infrequent clock changes, data is lost. Just like what would happen during some EM interference event.

Not particularly happy with this, and I'll have to take it into account when I implement the time sync. Eg: that any change larger then a few ms should not just update the PLL, but result in an immediate re-set of the clock. (And purge of the TX/RX buffers or adjustment of the deadlines/timestamps in them by the clock change.) Or something like that. In any case, complications indeed.

pavel-kirienko replied:
Maybe the actual root of the problem is that GNU/Linux is the wrong OS for real-time applications. Whether you timestamp in kernel space or in the user space (being able to use whatever time system you prefer), you get no guaranteed accuracy or latency bounds either way.

pavel-kirienko reviewed (changes requested):
Nice work. Please fix the tests.

TomDeRybel commented:
Hello Pavel,
I've updated the tests as you asked, and added a new one. There is a question there as well.
Other than that, I've made all the changes as I understand them.
Best regards,
Tom

pavel-kirienko commented on socketcan/libcanard/test/test_io.cpp:
```diff
@@ -85,28 +86,71 @@ TEST_CASE("IO-FD")
+    REQUIRE(1 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000, &loopback));  // Receive loopback frame on sa.
+    REQUIRE(loopback == true);
+    REQUIRE(fr.timestamp_usec > 0);
+    REQUIRE(fr.timestamp_usec >= old_ts); // TODO: time relation still sensible? Appears to be equal...
```
This seems right, yes. They may appear equal depending on the resolution of the timestamping clock.

pavel-kirienko commented on socketcan/libcanard/src/socketcan.h:
```diff
@@ -12,6 +12,15 @@
 /// --------------------------------------------------------------------------------------------------------------------
+/// Changelog
+///
+/// v0.1 - Added loop-back functionality.
+///        API change in socketcanPop(): loopback flag added.
+///      - Changed to kernel-based time-stamping for received frames for improved accuracy.
+///        API change in socketcanPop(): time stamp clock source is now CLOCK_REALTIME, vs CLOCK_TAI before.
+///
+/// v0.0 - Initial release
```
```suggestion
/// v2.0 - Added loop-back functionality.
///        API change in socketcanPop(): loopback flag added.
///      - Changed to kernel-based time-stamping for received frames for improved accuracy.
///        API change in socketcanPop(): time stamp clock source is now CLOCK_REALTIME, vs CLOCK_TAI before.
///
/// v1.0 - Initial release
```

pavel-kirienko replied:
```suggestion
/// v2.0 - Added loop-back functionality.
///        API change in socketcanPop(): loopback flag added.
///      - Changed to kernel-based time-stamping for received frames for improved accuracy.
///        API change in socketcanPop(): time stamp clock source is now CLOCK_REALTIME, vs CLOCK_TAI before.
///
/// v1.0 - Initial release
```
